pkg/types/openstack/validation/machinepool.go

```diff
@@ -25,5 +25,34 @@ func ValidateMachinePool(_ *openstack.Platform, machinePool *openstack.MachinePo
 		errs = append(errs, field.NotSupported(fldPath.Child("serverGroupPolicy"), machinePool.ServerGroupPolicy, validServerGroupPolicies))
 	}
 
+	if machinePool.RootVolume != nil {
+		rootVolumeType := machinePool.RootVolume.DeprecatedType
+		rootVolumeTypes := machinePool.RootVolume.Types
+		typePath := fldPath.Child("rootVolume").Child("type")
+		typesPath := fldPath.Child("rootVolume").Child("types")
+
+		if rootVolumeType != "" && len(rootVolumeTypes) > 0 {
+			errs = append(errs, field.Invalid(typePath, rootVolumeType, "Only one of type or types can be specified"))
+			errs = append(errs, field.Invalid(typesPath, rootVolumeTypes, "Only one of type or types can be specified"))
+		}
+
+		if rootVolumeType == "" && len(rootVolumeTypes) == 0 {
+			errs = append(errs, field.Invalid(typePath, rootVolumeType, "Either type or types must be specified"))
+			errs = append(errs, field.Invalid(typesPath, rootVolumeTypes, "Either type or types must be specified"))
+		}
+
+		// When distributing the Root volumes across multiple failure domains, we must use multiple Storage types so they use a different backend.
+		// Storage availability zones are purely cosmetic for now and can also be used to define where a volume should be created, however
+		// we don't want to force a user to define a Storage availability zone when using multiple Storage types.
+		// Therefore we decided to require as many Storage types as there are Compute availability zones, if there are multiple Compute availability zones.
+		// e.g. it would not make sense to have 3 Compute availability zones and 2 Storage types, because one of the Storage types would be used twice and
+		// therefore the number of failure domains would not be 3 anymore.
+		if machinePool.RootVolume.Types != nil {
+			if computes, volumes := len(machinePool.Zones), len(machinePool.RootVolume.Types); computes > 1 && volumes != computes {
```

Comment:
I've gave more thoughts on this one and this might be too opinionated.
I think the case of:

* 3 compute AZs
* no storage AZ
* one volume type (defined in `types`)

is a valid scenario (not ideal of course, since the root volume backend would be shared across the 3 failure domains...

Reply:
I have extended the logic (all lens must either be 0, 1, or equal to the highest) to check the volume type. PTAL